File: frontend/types/dataset.ts
export interface ColumnMeta {
  name: string
  dtype: string
  detected_role: string
  sample_values: unknown[]
}

export interface DatasetSummary {
  row_count: number
  column_count: number
  columns: string[]
  dtypes: Record<string, string>
  missing_values: Record<string, number>
  missing_pct: Record<string, number>
  numeric_columns: string[]
  categorical_columns: string[]
  date_columns: string[]
  memory_usage_mb: number
}

export interface Dataset {
  id: string
  upload_id: string
  name: string
  row_count: number
  column_count: number
  columns_meta: ColumnMeta[]
  summary: DatasetSummary
  file_path: string
  status: string
  error_msg: string | null
  created_at: string
}

export interface DatasetListResponse {
  datasets: Dataset[]
  total: number
}

export interface UploadResponse {
  id: string
  filename: string
  file_type: string
  file_size: number
  row_count: number
  column_count: number
  status: string
  error_msg: string | null
  created_at: string
  dataset_id?: string
}

export interface PreviewResponse {
  columns: string[]
  rows: Record<string, unknown>[]
  total_rows: number
  preview_rows: number
}

File: backend/app/schemas/dataset.py
from datetime import datetime
from typing import Optional

from pydantic import BaseModel


class ColumnMeta(BaseModel):
    name: str
    dtype: str
    detected_role: str | None = None
    sample_values: list | None = None


class DatasetResponse(BaseModel):
    id: str
    upload_id: str
    name: str
    row_count: int
    column_count: int
    columns_meta: list[ColumnMeta] | None = None
    summary: dict | None = None
    status: str
    error_msg: str | None = None
    file_path: str | None = None
    created_at: datetime

    class Config:
        from_attributes = True


class DatasetListResponse(BaseModel):
    datasets: list[DatasetResponse]
    total: int


class PreviewResponse(BaseModel):
    columns: list[str]
    rows: list[dict]
    total_rows: int
